GPII-2564: Add support for running Chrome in "headless" mode...

the-t-in-rtf opened the pull request:
See [GPII-2564](https://issues.gpii.net/browse/GPII-2564) for details.

amb26 commented:
This pull looks very promising and lot of GPU-related error cruft is gone from the console. However, there is a new error on every request (harmless-looking)

but also two of the tests run outrageously slowly. One of them, the remote fetch test, takes around 15 seconds, and another, "getCurrentUrl" usually takes over a minute. Here is a transcript of the middle section:

```
21:50:16.803:  jq: Test concluded - Module "chrome: Testing the driver's `get` function (local)..." Test name "chrome: Retrieve a local page...": 1/1 passed - P
ASS
[1009/215017.047:ERROR:devtools_http_handler.cc(786)]
DevTools listening on 127.0.0.1:12938

21:50:22.183:  Test case listener has not responded after 5000ms - at sequence pos 4 of 7 sequence element {
    "event": "{testEnvironment}.webdriver.events.onGetComplete",
    "listener": "{testEnvironment}.webdriver.getPageSource",
    "args": []
} of fixture chrome: Retrieve a remote page (requires network access)...
21:50:38.002:  Successfully queued test chrome: Get the browser's capabilities...
21:50:38.003:  jq: Test concluded - Module "chrome: Testing the driver's `get` function (remote)..." Test name "chrome: Retrieve a remote page (requires network
 access)...": 1/1 passed - PASS
[1009/215038.220:ERROR:devtools_http_handler.cc(786)]
DevTools listening on 127.0.0.1:12391

21:50:38.403:  Successfully queued test chrome: Get the current URL...
21:50:38.404:  jq: Test concluded - Module "chrome: Test `getCapabilities`..." Test name "chrome: Get the browser's capabilities...": 1/1 passed - PASS
[1009/215038.562:ERROR:devtools_http_handler.cc(786)]
DevTools listening on 127.0.0.1:12964

21:50:43.763:  Test case listener has not responded after 5000ms - at sequence pos 4 of 7 sequence element {
    "event": "{testEnvironment}.webdriver.events.onGetComplete",
    "listener": "{testEnvironment}.webdriver.getCurrentUrl",
    "args": []
} of fixture chrome: Get the current URL...
[1009/215156.240:INFO:CONSOLE(1)] "Uncaught undefined", source: https://tpc.googlesyndication.com/pagead/js/r20171004/r20110914/abg.js (1)
21:52:15.740:  Successfully queued test chrome: Retrieve a local page...
21:52:15.744:  jq: Test concluded - Module "chrome: Testing the driver's `getCurrentUrl` function..." Test name "chrome: Get the current URL...": 1/1 passed - P
ASS
```

The remote URL test appears to fetch http://www.google.com/ncr - which some back of the envelope research suggests is a feature which Google withdrew in 2015. The "getCurrentUrl" test loads http://caniuse.com which I can't see any clear reason to load very slowly but I notice that it does attempt to do a redirect.

the-t-in-rtf commented:
I suspect that we may be seeing the same kind of headless mode delays as we saw with gpii-testem.  I will add a `HEADLESS` flag and made "headed" the default.

the-t-in-rtf commented:
Although useful, the "headless" toggle doesn't seem to make much of a difference for the two slower tests.  I confirmed that caniuse.com takes ~30 seconds to load in the `getCurrentUrl` test in either mode.  I picked another URL and confirmed that the test completes in ~3 seconds.  I also updated the URL for the `get` test to avoid the redirect, that test also completes more quickly now.

@amb26, this should be ready for further review.

the-t-in-rtf commented:
This now also includes the work to update this package to test with Node LTS and npm@5, see [GPII-2832](https://issues.gpii.net/browse/GPII-2832) for details.

the-t-in-rtf commented:
This pull now includes the work required to updated all dependencies, including a version of infusion released after November 2017.

amb26 commented:
ok to test

amb26 commented:
ok to test

amb26 commented on src/js/webdriver.js:
```diff
@@ -83,9 +87,9 @@ gpii.webdriver.execute = function (that, fnName, eventName, fnArgs) {
- * @param that - The component itself
- * @param args {Array} - An array representing a series of function names and arguments.  Each array's first element is a function name. The remaining arguments are passed to the function.
- * @returns a Promise that will be resolved when navigation is complete, or rejected if there is an error.
+ * @param {Object} that - The component itself
```
For all "the component itself" please add the grade name rather than {Object}

the-t-in-rtf replied:
Done, throughout.

amb26 commented on src/js/webdriver.js:
```diff
@@ -112,9 +116,9 @@ gpii.webdriver.navigateHelper = function (that, args) {
- * @param actions {Object} - An array of action definitions.  Each element is an object with a `functionName` and `args` element.
- * @returns a Promise that will be resolved when the actions are complete, or rejected if there is an error.
+ * @param {Object} that - The component itself
+ * @param {Object} actionDefs - An array of action definitions.  Each element is an object with a `functionName` and `args` element.
```
Looks like this is more accurately Array<Something>

the-t-in-rtf replied:
Good catch.  I went further and added a `@typedef` as well and added other recent improved conventions (mentioning `fluid.promise` in return descriptions, et cetera).

amb26 commented on tests/js/lib/globalErrorHandler.js:
```diff
@@ -0,0 +1,25 @@
+    }
+});
+
+fluid.registerNamespace("gpii.test.webdriver");
```
This is unnecessary since you just issued the defaults above

the-t-in-rtf replied:
Removed.

amb26 commented:
I tried testing this with a deliberately mismatched version of chromedriver and chrome and got what looked like a lot of fatal errors which were reported in a rather mangled way and in addition did not result in test failures. Here is a sample:
````
20:07:27.595:  ASSERTION FAILED: {

}
20:07:27.596:  Current activity:
    while firing to listener to event named onError of component {
    "typeName": "gpii.webdriver",
    "id": "7bq56c6p-333",
    "lifecycleStatus": "treeConstructed",
    "destroy": { Function

    },
    "options": {
        "gradeNames": [
            "fluid.component",
            "gpii.webdriver"
        ],
        "browser": "chrome",
        "events": {
            "onDriverReady": "{testEnvironment}.events.onDriverReady",
            "stopFixtures": "{testEnvironment}.events.stopFixtures",
            "onActionsHelperComplete": null,
            "onDumpLogsComplete": null,
            "onError": null,
            "onNavigateHelperComplete": null,
            "onActionsComplete": null,
            "onCallComplete": null,
            "onCloseComplete": null,
            "onExecuteAsyncScriptComplete": null,
            "onExecuteScriptComplete": null,
            "onFindElementComplete": null,
            "onFindElementsComplete": null,
            "onGetComplete": null,
            "onGetAllWindowHandlesComplete": null,
            "onGetCapabilitiesComplet .... [output suppressed at 1024 chars - for more output, increase fluid.logObjectRenderChars]
    while invoking invoker with name actionsHelper and record {
    "func": undefined,
    "funcName": "gpii.webdriver.actionsHelper",
    "listener": undefined,
    "this": undefined,
    "method": undefined,
    "changePath": undefined,
    "value": undefined,
    "args": [
        "{that}",
        "{arguments}.0"
    ],
    "componentSource": "gpii.webdriver"
} from path chrome-7bq56c6p-297,webdriver holding component {
    "typeName": "gpii.webdriver",
    "id": "7bq56c6p-333",
    "lifecycleStatus": "treeConstructed",
    "destroy": { Function

    },
    "options": {
        "gradeNames": [
            "fluid.component",
            "gpii.webdriver"
        ],
        "browser": "chrome",
        "events": {
            "onDriverReady": "{testEnvironment}.events.onDriverReady",
            "stopFixtures": "{testEnvironment}.events.stopFixtures",
            "onActionsHelperComplete": null,
            "onDumpLogsComplete": null,
            "onError": null,
            "onNavigateHelperComplete": null,
            "onActionsComplete": null,
            "onCallComplete": null,
            "onCloseComplete": null,
            "onExecuteAsyncScriptComplete": null,
            "onExecuteScriptComplete": null,
            "onFindElementComplete": null,
            "onFindElementsComplete": null,
            "onGetComplete": null,
            "onGetAllWindowHandlesComplete": null,
            "onGetCapabilitiesComplet .... [output suppressed at 1024 chars - for more output, increase fluid.logObjectRenderChars]
20:07:27.609:  ASSERTION FAILED: {
    "message": "Assertion failure - check console for more details: {\n    \n}",
    "stack": "Error: Assertion failure - check console for more details: {\n    \n}\n    at new fluid.FluidError (E:\\source\\gits\\gpii-web
driver\\node_modules\\infusion\\src\\framework\\core\\js\\Fluid.js:179:26)\n    at Object.fluid.builtinFail (E:\\source\\gits\\gpii-webdrive
r\\node_modules\\infusion\\src\\framework\\core\\js\\Fluid.js:203:15)\n    at kettle.failureHandler (E:\\source\\gits\\gpii-webdriver\\node_
modules\\kettle\\lib\\KettleUtils.js:60:11)\n    at Object.fire (E:\\source\\gits\\gpii-webdriver\\node_modules\\infusion\\src\\framework\\c
ore\\js\\Fluid.js:1568:40)\n    at fluid.fail (E:\\source\\gits\\gpii-webdriver\\node_modules\\infusion\\src\\framework\\core\\js\\Fluid.js:
218:32)\n    at togo (E:\\source\\gits\\gpii-webdriver\\node_modules\\infusion\\src\\framework\\core\\js\\FluidIoC.js:1806:33)\n    at fire
(E:\\source\\gits\\gpii-webdriver\\node_modules\\infusion\\src\\framework\\core\\js\\Fluid.js:1568:40 .... [output suppressed at 1024 chars
- for more output, increase fluid.logObjectRenderChars]
20:07:27.611:  Current activity:
    while firing to listener to event named onError of component {
    "typeName": "gpii.webdriver",
    "id": "7bq56c6p-333",
    "lifecycleStatus": "treeConstructed",
    "destroy": { Function

    },
    "options": {
        "gradeNames": [
            "fluid.component",
            "gpii.webdriver"
        ],
        "browser": "chrome",
        "events": {
            "onDriverReady": "{testEnvironment}.events.onDriverReady",
            "stopFixtures": "{testEnvironment}.events.stopFixtures",
            "onActionsHelperComplete": null,
            "onDumpLogsComplete": null,
            "onError": null,
            "onNavigateHelperComplete": null,
            "onActionsComplete": null,
            "onCallComplete": null,
            "onCloseComplete": null,
            "onExecuteAsyncScriptComplete": null,
            "onExecuteScriptComplete": null,
            "onFindElementComplete": null,
            "onFindElementsComplete": null,
            "onGetComplete": null,
            "onGetAllWindowHandlesComplete": null,
            "onGetCapabilitiesComplet .... [output suppressed at 1024 chars - for more output, increase fluid.logObjectRenderChars]
20:07:27.615:  ASSERTION FAILED: {
    "message": "Assertion failure - check console for more details: {\n    \"message\": \"Assertion failure - check console for more details
: {\\n    \\n}\",\n    \"stack\": \"Error: Assertion failure - check console for more details: {\\n    \\n}\\n    at new fluid.FluidError (E
:\\\\source\\\\gits\\\\gpii-webdriver\\\\node_modules\\\\infusion\\\\src\\\\framework\\\\core\\\\js\\\\Fluid.js:179:26)\\n    at Object.flui
d.builtinFail (E:\\\\source\\\\gits\\\\gpii-webdriver\\\\node_modules\\\\infusion\\\\src\\\\framework\\\\core\\\\js\\\\Fluid.js:203:15)\\n
  at kettle.failureHandler (E:\\\\source\\\\gits\\\\gpii-webdriver\\\\node_modules\\\\kettle\\\\lib\\\\KettleUtils.js:60:11)\\n    at Object
.fire (E:\\\\source\\\\gits\\\\gpii-webdriver\\\\node_modules\\\\infusion\\\\src\\\\framework\\\\core\\\\js\\\\Fluid.js:1568:40)\\n    at fl
uid.fail (E:\\\\source\\\\gits\\\\gpii-webdriver\\\\node_modules\\\\infusion\\\\src\\\\framework\\\\core\\\\js\\\\Fluid.js:218:32)\\n    at
togo (E:\\\\source\\\\gits\\\\gpii-webdriv .... [output suppressed at 1024 chars - for more output, increase fluid.logObjectRenderChars]
20:07:27.617:  Current activity:
    while firing to listener to event named onError of component {
    "typeName": "gpii.webdriver",
    "id": "7bq56c6p-333",
    "lifecycleStatus": "treeConstructed",
    "destroy": { Function

    },
    "options": {
        "gradeNames": [
            "fluid.component",
            "gpii.webdriver"
        ],
        "browser": "chrome",
        "events": {
            "onDriverReady": "{testEnvironment}.events.onDriverReady",
            "stopFixtures": "{testEnvironment}.events.stopFixtures",
            "onActionsHelperComplete": null,
            "onDumpLogsComplete": null,
            "onError": null,
            "onNavigateHelperComplete": null,
            "onActionsComplete": null,
            "onCallComplete": null,
            "onCloseComplete": null,
            "onExecuteAsyncScriptComplete": null,
            "onExecuteScriptComplete": null,
            "onFindElementComplete": null,
            "onFindElementsComplete": null,
            "onGetComplete": null,
            "onGetAllWindowHandlesComplete": null,
            "onGetCapabilitiesComplet .... [output suppressed at 1024 chars - for more output, increase fluid.logObjectRenderChars]
20:07:27.629:  FATAL ERROR: Uncaught exception: Assertion failure - check console for more details: {
    "message": "Assertion failure - check console for more details: {\n    \"message\": \"Assertion failure - check console for more details
: {\\n    \\n}\",\n    \"stack\": \"Error: Assertion failure - check console for more details: {\\n    \\n}\\n    at new fluid.FluidError (E
:\\\\source\\\\gits\\\\gpii-webdriver\\\\node_modules\\\\infusion\\\\src\\\\framework\\\\core\\\\js\\\\Fluid.js:179:26)\\n    at Object.flui
d.builtinFail (E:\\\\source\\\\gits\\\\gpii-webdriver\\\\node_modules\\\\infusion\\\\src\\\\framework\\\\core\\\\js\\\\Fluid.js:203:15)\\n
  at kettle.failureHandler (E:\\\\source\\\\gits\\\\gpii-webdriver\\\\node_modules\\\\kettle\\\\lib\\\\KettleUtils.js:60:11)\\n    at Object
.fire (E:\\\\source\\\\gits\\\\gpii-webdriver\\\\node_modules\\\\infusion\\\\src\\\\framework\\\\core\\\\js\\\\Fluid.js:1568:40)\\n    at fl
uid.fail (E:\\\\source\\\\gits\\\\gpii-webdriver\\\\node_modules\\\\infusion\\\\src\\\\framework\ .... [output suppressed at 1024 chars - fo
r more output, increase fluid.logObjectRenderChars]
Error: Assertion failure - check console for more details: {
    "message": "Assertion failure - check console for more details: {\n    \"message\": \"Assertion failure - check console for more details
: {\\n    \\n}\",\n    \"stack\": \"Error: Assertion failure - check console for more details: {\\n    \\n}\\n    at new fluid.FluidError (E
:\\\\source\\\\gits\\\\gpii-webdriver\\\\node_modules\\\\infusion\\\\src\\\\framework\\\\core\\\\js\\\\Fluid.js:179:26)\\n    at Object.flui
d.builtinFail (E:\\\\source\\\\gits\\\\gpii-webdriver\\\\node_modules\\\\infusion\\\\src\\\\framework\\\\core\\\\js\\\\Fluid.js:203:15)\\n
  at kettle.failureHandler (E:\\\\source\\\\gits\\\\gpii-webdriver\\\\node_modules\\\\kettle\\\\lib\\\\KettleUtils.js:60:11)\\n    at Object
.fire (E:\\\\source\\\\gits\\\\gpii-webdriver\\\\node_modules\\\\infusion\\\\src\\\\framework\\\\core\\\\js\\\\Fluid.js:1568:40)\\n    at fl
uid.fail (E:\\\\source\\\\gits\\\\gpii-webdriver\\\\node_modules\\\\infusion\\\\src\\\\framework\\\\core\\\\js\\\\Fluid.js:218:32)\\n    at
togo (E:\\\\source\\\\gits\\\\gpii-webdriv .... [output suppressed at 1024 chars - for more output, increase fluid.logObjectRenderChars]
    at new fluid.FluidError (E:\source\gits\gpii-webdriver\node_modules\infusion\src\framework\core\js\Fluid.js:179:26)
    at Object.fluid.builtinFail (E:\source\gits\gpii-webdriver\node_modules\infusion\src\framework\core\js\Fluid.js:203:15)
    at kettle.failureHandler (E:\source\gits\gpii-webdriver\node_modules\kettle\lib\KettleUtils.js:60:11)
    at Object.fire (E:\source\gits\gpii-webdriver\node_modules\infusion\src\framework\core\js\Fluid.js:1568:40)
    at fluid.fail (E:\source\gits\gpii-webdriver\node_modules\infusion\src\framework\core\js\Fluid.js:218:32)
    at togo (E:\source\gits\gpii-webdriver\node_modules\infusion\src\framework\core\js\FluidIoC.js:1806:33)
    at fire (E:\source\gits\gpii-webdriver\node_modules\infusion\src\framework\core\js\Fluid.js:1568:40)
    at ManagedPromise.invokeCallback_ (E:\source\gits\gpii-webdriver\node_modules\selenium-webdriver\lib\promise.js:1376:14)
    at TaskQueue.execute_ (E:\source\gits\gpii-webdriver\node_modules\selenium-webdriver\lib\promise.js:3084:14)
    at TaskQueue.executeNext_ (E:\source\gits\gpii-webdriver\node_modules\selenium-webdriver\lib\promise.js:3067:27)
    at asyncRun (E:\source\gits\gpii-webdriver\node_modules\selenium-webdriver\lib\promise.js:2927:27)
    at E:\source\gits\gpii-webdriver\node_modules\selenium-webdriver\lib\promise.js:668:7
    at process._tickCallback (internal/process/next_tick.js:68:7)
20:07:27.635:  Current activity:
20:07:27.728:  Successfully queued test chrome: Use the tab key to navigate to a particular field...
````

I think an important step is to ensure that your gpii.webdriver.loadTestingSupport also makes an invocation of kettle.loadTestingSupport without which a lot of quite strange things will happen, including the failure to escalate uncaught exceptions into test failures. Also you should look carefully at the structure of the payload sent to your onError event to see if an intelligible cause of the failure can be extracted from it rather than slinging the whole structure to fluid.fail

the-t-in-rtf commented:
> I think an important step is to ensure that your gpii.webdriver.loadTestingSupport also makes an invocation of kettle.loadTestingSupport without which a lot of quite strange things will happen, including the failure to escalate uncaught exceptions into test failures.

Good catch, and fixed.

> Also you should look carefully at the structure of the payload sent to your onError event to see if an intelligible cause of the failure can be extracted from it rather than slinging the whole structure to fluid.fail

I added a new function that looks for `message` (a common convention throughout the infusion/kettle ecosystem) as well as a few other variations I tend to use.

the-t-in-rtf commented:
@amb26, this should be ready for further review.

amb26 commented:
Errors are now better formatted but are still not resulting in test failures. Here is a sample:
````
13:09:46.814:  ASSERTION FAILED: Promise resolver undefined is not a function
13:09:46.815:  Current activity:
13:09:46.816:      while firing to listener to event named onError of component {
    "typeName": "gpii.webdriver",
    "id": "2wnaoy20-382",
    "lifecycleStatus": "treeConstructed",
    "destroy": { Function

    },
    "options": {
        "gradeNames": [
            "fluid.component",
            "gpii.webdriver"
        ],
        "events": {
            "onDriverReady": "{testEnvironment}.events.onDriverReady",
            "stopFixtures": "{testEnvironment}.events.stopFixtures",
            "onActionsHelperComplete": null,
            "onDumpLogsComplete": null,
            "onError": null,
            "onNavigateHelperComplete": null,
            "onActionsComplete": null,
            "onCallComplete": null,
            "onCloseComplete": null,
            "onExecuteAsyncScriptComplete": null,
            "onExecuteScriptComplete": null,
            "onFindElementComplete": null,
            "onFindElementsComplete": null,
            "onGetComplete": null,
            "onGetAllWindowHandlesComplete": null,
            "onGetCapabilitiesComplete": null,
            "onGetC .... [output suppressed at 1024 chars - for more output, increase fluid.logObjectRenderChars]
13:09:46.821:      while invoking invoker with name actionsHelper and record {
    "func": undefined,
    "funcName": "gpii.webdriver.actionsHelper",
    "listener": undefined,
    "this": undefined,
    "method": undefined,
    "changePath": undefined,
    "value": undefined,
    "args": [
        "{that}",
        "{arguments}.0"
    ],
    "componentSource": "gpii.webdriver"
} from path chrome-2wnaoy20-343.webdriver holding component {
    "typeName": "gpii.webdriver",
    "id": "2wnaoy20-382",
    "lifecycleStatus": "treeConstructed",
    "destroy": { Function

    },
    "options": {
        "gradeNames": [
            "fluid.component",
            "gpii.webdriver"
        ],
        "events": {
            "onDriverReady": "{testEnvironment}.events.onDriverReady",
            "stopFixtures": "{testEnvironment}.events.stopFixtures",
            "onActionsHelperComplete": null,
            "onDumpLogsComplete": null,
            "onError": null,
            "onNavigateHelperComplete": null,
            "onActionsComplete": null,
            "onCallComplete": null,
            "onCloseComplete": null,
            "onExecuteAsyncScriptComplete": null,
            "onExecuteScriptComplete": null,
            "onFindElementComplete": null,
            "onFindElementsComplete": null,
            "onGetComplete": null,
            "onGetAllWindowHandlesComplete": null,
            "onGetCapabilitiesComplete": null,
            "onGetC .... [output suppressed at 1024 chars - for more output, increase fluid.logObjectRenderChars]
13:09:46.835:  ASSERTION FAILED: Assertion failure - check console for more details: Promise resolver undefined is not a function
13:09:46.836:  Current activity:
13:09:46.836:      while firing to listener to event named onError of component {
    "typeName": "gpii.webdriver",
    "id": "2wnaoy20-382",
    "lifecycleStatus": "treeConstructed",
    "destroy": { Function

    },
    "options": {
        "gradeNames": [
            "fluid.component",
            "gpii.webdriver"
        ],
        "events": {
            "onDriverReady": "{testEnvironment}.events.onDriverReady",
            "stopFixtures": "{testEnvironment}.events.stopFixtures",
            "onActionsHelperComplete": null,
            "onDumpLogsComplete": null,
            "onError": null,
            "onNavigateHelperComplete": null,
            "onActionsComplete": null,
            "onCallComplete": null,
            "onCloseComplete": null,
            "onExecuteAsyncScriptComplete": null,
            "onExecuteScriptComplete": null,
            "onFindElementComplete": null,
            "onFindElementsComplete": null,
            "onGetComplete": null,
            "onGetAllWindowHandlesComplete": null,
            "onGetCapabilitiesComplete": null,
            "onGetC .... [output suppressed at 1024 chars - for more output, increase fluid.logObjectRenderChars]
13:09:46.841:  ASSERTION FAILED: Assertion failure - check console for more details: Assertion failure - check console for more details:
````
Possibly a missing include as well?
Here's a failure from lower down:
````
13:11:21.846:  ASSERTION FAILED: Custom Message.
Wait timed out after 250ms
13:11:21.847:  Current activity:
13:11:21.848:      while firing to listener to event named onError of component {
    "typeName": "gpii.webdriver",
    "id": "2wnaoy20-3221",
    "lifecycleStatus": "treeConstructed",
    "destroy": { Function

    },
    "options": {
        "gradeNames": [
            "fluid.component",
            "gpii.webdriver"
        ],
        "events": {
            "onDriverReady": "{testEnvironment}.events.onDriverReady",
            "stopFixtures": "{testEnvironment}.events.stopFixtures",
            "onActionsHelperComplete": null,
            "onDumpLogsComplete": null,
            "onError": null,
            "onNavigateHelperComplete": null,
            "onActionsComplete": null,
            "onCallComplete": null,
            "onCloseComplete": null,
            "onExecuteAsyncScriptComplete": null,
            "onExecuteScriptComplete": null,
            "onFindElementComplete": null,
            "onFindElementsComplete": null,
            "onGetComplete": null,
            "onGetAllWindowHandlesComplete": null,
            "onGetCapabilitiesComplete": null,
            "onGet .... [output suppressed at 1024 chars - for more output, increase fluid.logObjectRenderChars]
13:11:21.854:  ASSERTION FAILED: Assertion failure - check console for more details: Custom Message.
Wait timed out after 250ms
13:11:21.854:  Current activity:
13:11:21.855:      while firing to listener to event named onError of component {
    "typeName": "gpii.webdriver",
    "id": "2wnaoy20-3221",
    "lifecycleStatus": "treeConstructed",
    "destroy": { Function

    },
    "options": {
        "gradeNames": [
            "fluid.component",
            "gpii.webdriver"
        ],
        "events": {
            "onDriverReady": "{testEnvironment}.events.onDriverReady",
            "stopFixtures": "{testEnvironment}.events.stopFixtures",
            "onActionsHelperComplete": null,
            "onDumpLogsComplete": null,
            "onError": null,
            "onNavigateHelperComplete": null,
            "onActionsComplete": null,
            "onCallComplete": null,
            "onCloseComplete": null,
            "onExecuteAsyncScriptComplete": null,
            "onExecuteScriptComplete": null,
            "onFindElementComplete": null,
            "onFindElementsComplete": null,
            "onGetComplete": null,
            "onGetAllWindowHandlesComplete": null,
            "onGetCapabilitiesComplete": null,
            "onGet .... [output suppressed at 1024 chars - for more output, increase fluid.logObjectRenderChars]
13:11:21.858:  ASSERTION FAILED: Assertion failure - check console for more details: Assertion failure - check console for more details: Cus
tom Message.
Wait timed out after 250ms
13:11:21.859:  Current activity:
13:11:21.859:      while firing to listener to event named onError of component {
    "typeName": "gpii.webdriver",
    "id": "2wnaoy20-3221",
    "lifecycleStatus": "treeConstructed",
    "destroy": { Function

    },
    "options": {
        "gradeNames": [
            "fluid.component",
            "gpii.webdriver"
        ],
        "events": {
            "onDriverReady": "{testEnvironment}.events.onDriverReady",
            "stopFixtures": "{testEnvironment}.events.stopFixtures",
            "onActionsHelperComplete": null,
            "onDumpLogsComplete": null,
            "onError": null,
            "onNavigateHelperComplete": null,
            "onActionsComplete": null,
            "onCallComplete": null,
            "onCloseComplete": null,
            "onExecuteAsyncScriptComplete": null,
            "onExecuteScriptComplete": null,
            "onFindElementComplete": null,
            "onFindElementsComplete": null,
            "onGetComplete": null,
            "onGetAllWindowHandlesComplete": null,
            "onGetCapabilitiesComplete": null,
            "onGet .... [output suppressed at 1024 chars - for more output, increase fluid.logObjectRenderChars]
13:11:21.863:  ASSERTION FAILED: Assertion failure - check console for more details: Assertion failure - check console for more details: Ass
ertion failure - check console for more details: Custom Message.
Wait timed out after 250ms
13:11:21.864:  Current activity:
13:11:21.865:      while firing to listener to event named onError of component {
    "typeName": "gpii.webdriver",
    "id": "2wnaoy20-3221",
    "lifecycleStatus": "treeConstructed",
    "destroy": { Function

    },
    "options": {
        "gradeNames": [
            "fluid.component",
            "gpii.webdriver"
        ],
        "events": {
            "onDriverReady": "{testEnvironment}.events.onDriverReady",
            "stopFixtures": "{testEnvironment}.events.stopFixtures",
            "onActionsHelperComplete": null,
            "onDumpLogsComplete": null,
            "onError": null,
            "onNavigateHelperComplete": null,
            "onActionsComplete": null,
            "onCallComplete": null,
            "onCloseComplete": null,
            "onExecuteAsyncScriptComplete": null,
            "onExecuteScriptComplete": null,
            "onFindElementComplete": null,
            "onFindElementsComplete": null,
            "onGetComplete": null,
            "onGetAllWindowHandlesComplete": null,
            "onGetCapabilitiesComplete": null,
            "onGet .... [output suppressed at 1024 chars - for more output, increase fluid.logObjectRenderChars]
13:11:21.871:  FATAL ERROR: Uncaught exception: Assertion failure - check console for more details: Assertion failure - check console for mo
re details: Assertion failure - check console for more details: Assertion failure - check console for more details: Custom Message.
Wait timed out after 250ms
Error: Assertion failure - check console for more details: Assertion failure - check console for more details: Assertion failure - check con
sole for more details: Assertion failure - check console for more details: Custom Message.
Wait timed out after 250ms
    at new fluid.FluidError (E:\source\gits\gpii-webdriver\node_modules\infusion\src\framework\core\js\Fluid.js:177:26)
    at Object.fluid.builtinFail (E:\source\gits\gpii-webdriver\node_modules\infusion\src\framework\core\js\Fluid.js:201:15)
    at kettle.failureHandler (E:\source\gits\gpii-webdriver\node_modules\kettle\lib\KettleUtils.js:60:11)
    at Object.fire (E:\source\gits\gpii-webdriver\node_modules\infusion\src\framework\core\js\Fluid.js:1663:40)
    at Object.fluid.fail (E:\source\gits\gpii-webdriver\node_modules\infusion\src\framework\core\js\Fluid.js:218:32)
    at gpii.webdriver.throwError (E:\source\gits\gpii-webdriver\src\js\webdriver.js:9:6634)
    at togo (E:\source\gits\gpii-webdriver\node_modules\infusion\src\framework\core\js\FluidIoC.js:2417:33)
    at fire (E:\source\gits\gpii-webdriver\node_modules\infusion\src\framework\core\js\Fluid.js:1663:40)
    at ManagedPromise.invokeCallback_ (E:\source\gits\gpii-webdriver\node_modules\selenium-webdriver\lib\promise.js:1376:14)
    at TaskQueue.execute_ (E:\source\gits\gpii-webdriver\node_modules\selenium-webdriver\lib\promise.js:3084:14)
    at TaskQueue.executeNext_ (E:\source\gits\gpii-webdriver\node_modules\selenium-webdriver\lib\promise.js:3067:27)
    at asyncRun (E:\source\gits\gpii-webdriver\node_modules\selenium-webdriver\lib\promise.js:2927:27)
    at E:\source\gits\gpii-webdriver\node_modules\selenium-webdriver\lib\promise.js:668:7
    at process._tickCallback (internal/process/next_tick.js:68:7)
````

the-t-in-rtf commented:
@amb26, as discussed on chat, the errors you mentioned are from failure tests.  As agreed, I've added a warning message that's logged right before an expected error is thrown.

the-t-in-rtf commented:
Thanks, @amb26, I have cut the 1.1.1 release as discussed on chat, and have updated the forward-facing version in master.
